### PR TITLE
Restore CUDA half linspace+logspace and add coverage tests

### PR DESCRIPTION
**Pull request by zou3519**
This PR restores the implementation of CUDA half linspace+logspace.

I added tests for the following:
- linspace+logspace have the same support for integral types on CPU/CUDA
- Precision tests for CUDA half, float, and double.

The precision for CUDA half seems bad, but I checked the numbers against
previous versions of pytorch. The output of CUDA Half linspace+logspace
are exactly the same when compared with 1.2.0.

Equivalent-ish PR on master:
https://github.com/pytorch/pytorch/pull/31962

### CODE REVIEW

File: test/test_torch.py [hunk 1]
```diff
@@ -13709,22 +13719,50 @@ def test_cat_big(self, device):
 
 # Tests that compare a device's computation with the (gold-standard) CPU's.
 class TestDevicePrecision(TestCase):
-    def test_linspace(self, device):
-        a = torch.linspace(0, 10, 10, device=device)
-        b = torch.linspace(0, 10, 10)
+    # The implementation of linspace+logspace have special cases steps=0 and steps=1
```
Comment (gchanan) [inline test/test_torch.py]: what special steps?

File: test/test_torch.py [hunk 2]
```diff
@@ -13709,22 +13719,50 @@ def test_cat_big(self, device):
...
+    def _test_linspace(self, device, dtype, steps):
+        a = torch.linspace(0, 10, steps=steps, dtype=dtype, device=device)
+        b = torch.linspace(0, 10, steps=steps)
         self.assertEqual(a, b)
 
-    @dtypes(torch.double)
-    def test_logspace(self, device, dtype):
-        a = torch.logspace(1, 10, 10, dtype=dtype, device=device)
-        b = torch.logspace(1, 10, 10, dtype=dtype, device='cpu')
+    @dtypes(torch.half, torch.float, torch.double)
```
Comment (gchanan) [inline test/test_torch.py]: does half work on CPU?  otherwise don't you need to do something like:
```
@dtypesIfCUDA(torch.half, torch.float, torch.double)
@dtypes(torch.float, torch.double)
```

File: test/test_torch.py (hunk 1 above)
Comment (gchanan) [reply test/test_torch.py]: can you just comment what's special about these?

File: test/test_torch.py [hunk 3]
```diff
@@ -13709,22 +13719,50 @@ def test_cat_big(self, device):
...
+    @precisionOverride({torch.half: 4e-3})
```
Comment (gchanan) [inline test/test_torch.py]: for all precision overrides can you write how you came up with the number?

File: test/test_torch.py [hunk 4]
```diff
@@ -10560,6 +10560,16 @@ def test_linspace(self, device):
         y = torch.linspace(0, 3, 4, out=x.narrow(1, 1, 2))
         self.assertEqual(x, torch.tensor(((0, 0, 1), (0, 2, 3)), device=device), 0)
 
+    def test_linspace_integral(self, device):
```
Comment (gchanan) [inline test/test_torch.py]: can you split this into another PR (or just remove this) so the PRs are basically the same on master and v1.4.0?  Otherwise it becomes a bit harder to track what is what.

**Review by gchanan — COMMENTED** [review]
can you check the precision on master?

**Review by gchanan — APPROVED** [review]
see note in master PR (https://github.com/pytorch/pytorch/pull/31962).

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit f4071be8:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=f4071be82a4adfd338d40dd85b9119758ea38886).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4175989) pytorch_macos_10_13_py3_test (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=82292769) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=82292769))
<details>
<summary>
<code>Jan 08 15:22:17 RuntimeError: test_fake_quant failed!</code>
</summary>

```
Jan 08 15:22:17 Ran 11 tests in 3.272s 
Jan 08 15:22:17  
Jan 08 15:22:17 FAILED (errors=3, skipped=1) 
Jan 08 15:22:17  
Jan 08 15:22:17 Generating XML reports... 
Jan 08 15:22:17 Traceback (most recent call last): 
Jan 08 15:22:17   File "test/run_test.py", line 455, in <module> 
Jan 08 15:22:17     main() 
Jan 08 15:22:17   File "test/run_test.py", line 448, in main 
Jan 08 15:22:17     raise RuntimeError(message) 
Jan 08 15:22:17 RuntimeError: test_fake_quant failed! 
Jan 08 15:22:17 + cleanup 
Jan 08 15:22:17 + retcode=1 
Jan 08 15:22:17 + set +x 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>
